kernel/sched.c

```diff
@@ -728,19 +746,29 @@ static void z_thread_halt(struct k_thread *thread, k_spinlock_key_t key,
 
 	if (is_halting(thread) && (thread != _current)) {
 		if (arch_is_in_isr()) {
+			_current_cpu->spinning_on = cpu;
+
 			/* ISRs can only spin waiting another CPU */
 			k_spin_unlock(&sched_spinlock, key);
 			while (is_halting(thread)) {
 			}
 
 			/* Now we know it's halting, but not necessarily
 			 * halted (suspended or aborted). Wait for the switch
-			 * to happen!
+			 * to happen unless deadlock was detected. If deadlock
+			 * was detected, do not spin and we will be the first
+			 * CPU of the chain of spinners to complete, and the
+			 * others will exit their spin loops in turn.
 			 */
 			key = k_spin_lock(&sched_spinlock);
-			z_sched_switch_spin(thread);
+
+			if ((cpu->current == thread) &&
```

Comment:
Pondering this some more. Maybe, just maybe there is a way out.

Reply:
If k_thread structures were static and never dynamic, then we could put a field in the k_thread that could be used to  block/delay a thread from re-using it until all CPUs trying to abort the thread finish. Unfortunately, this is a restriction that we can not assume, so scratch that.

We could ignore the re-use scenario. That would put us to no worse off than we are now. But that still leaves us with an identified problem/hole.

The best that I can think of is to add an unique 32-bit ID field to the k_thread. This field would get initialised based on a global variable that monotonically increases each time a thread is created. This should give us a high degree of confidence in being able to detect a re-used thread. Due to the finite width of this field, it would not be fool-proof. For it to fail, it would require about N * 2^32 (where N is a whole number greater than 0) to have been created since the original thread that we are trying to delete was created.